File: src/falcon_core/math/discrete_spaces/base_discrete_space.py
"""A discrete space can fill an entire domain with a set of discrete values."""

from typing import TYPE_CHECKING

from .dependancies import (
    Axes,
    Jsonable,
    LabelledControlArray,
)
from .typing import (
    ControlArray,
    Knob,
)

if TYPE_CHECKING:
    from .dependancies import Domain
    from .typing import (
        CoupledKnobDomain,
        UnitSpace,
    )


class BaseDiscreteSpace(Jsonable):
    """A discrete space can fill an entire domain with a set of discrete values."""

    _space: "UnitSpace"
    _axes: "Axes[CoupledKnobDomain]"

    def __init__(
        self,
        space: "UnitSpace",
        axes: "Axes[CoupledKnobDomain]",
    ):
        """Initialize the DiscreteSpace object.

        The order of the Knobs in the axes are defined to line up with the space.

        Args:
            space: the space that the discrete values fill.
            axes: the axes of the discrete values.
        """
        self._space = space
        self._axes = axes
        self.validate_unit_space_dimensionality_matches_knobs()
        self.validate_knob_uniqueness()

    def validate_unit_space_dimensionality_matches_knobs(self):
        """Validate that the unit space dimensionality matches the knobs."""
        assert self._space.dimension == self._axes.dimension, (
            "The unit space dimensionality must match the knobs."
        )

    def validate_knob_uniqueness(self):
        """Validate that the knobs are unique."""
        old_names = set()
        for axis in self._axes:
            new_names = set(axis.knobs.get_default_names())
            assert new_names.isdisjoint(old_names), "The default names must be unique."
            old_names.update(new_names)

    def get_axis(
        self,
        knob: "Knob",
    ) -> int:
        """Return the index of the axis containing the given knob.

        Args:
            knob: The knob to search for.

        Returns:
            The index of the axis containing the given knob.

        Raises:
            ValueError: If the knob is not found.
        """
        for axis in self._axes:
            if knob in axis.knobs:
                return self._axes.index(axis)
        msg = f"Knob {knob} not found in the axes."
        raise ValueError(msg)

    def get_domain(
        self,
        knob: "Knob",
    ) -> "Domain":
        """Return the domain of the given knob.

        Args:
            knob: The knob to search for.

        Returns:
            The domain of the given knob.
        """
        axis = self.get_axis(knob=knob)
        coupled_domain = self._axes[axis]
        return coupled_domain.get_domain(search=knob)

    def get_projection(
        self,
        projection: "Axes[Knob]",
    ) -> "Axes[LabelledControlArray]":
        """Return the projection of the space onto the given axes.

        Args:
            projection: The axes to project onto.

        Returns:
            The projection of the space onto the given axes.
        """
        assert projection.dimension == self._space.dimension, (
            "The projection dimensionality must be less than or equal to the space dimensionality."
        )
        projection_axes = Axes([self.get_axis(knob=knob) for knob in projection])

        assert projection.dimension == len(set(projection_axes.axes)), (
            "The projection axes must be unique."
        )
        unitprojections = self._space.create_array(axes=projection_axes)

        scaled_projections: list[ControlArray] = []
        for unitprojection, knob in zip(unitprojections, projection):
            assert isinstance(unitprojection, ControlArray), (
                f"Expected the type of each unit projection to be Control Array, but it was {type(unitprojection)}"
            )
            assert isinstance(knob, Knob), (
                f"Expected the type of the the knob to be a Knob, but it was {type(knob)}"
            )
            domain = self.get_domain(knob=knob)
            difference = domain.range
            min_value = domain.lesser_bound
            scaled_projections.append(unitprojection * difference + min_value)

        return Axes(
            [
                LabelledControlArray.from_port(array=proj.data, port=knob)
                for proj, knob in zip(scaled_projections, projection)
            ]
        )

File: src/math/discrete_spaces/BaseDiscreteSpace.cpp
#include "falcon_core/math/discrete_spaces/BaseDiscreteSpace.hpp"

#include <stdexcept>

#include "falcon_core/instrument_interfaces/names/InstrumentPort.hpp"
#include "falcon_core/math/domains/CoupledLabelledDomain.hpp"

namespace falcon_core::math::discrete_spaces {

BaseDiscreteSpace::BaseDiscreteSpace() = default;
BaseDiscreteSpace::BaseDiscreteSpace(
    const spaces::UnitSpaceSP&                     space,
    const AxesSP<domains::CoupledLabelledDomain>&  axes,
    const AxesSP<generic::Map<std::string, bool>>& increasing)
    : _space(space), _axes(axes), _increasing(increasing) {
  validate_unit_space_dimensionality_matches_knobs();
  validate_knob_uniqueness();
}
const spaces::UnitSpaceSP& BaseDiscreteSpace::space() const { return _space; }
const AxesSP<domains::CoupledLabelledDomain>& BaseDiscreteSpace::axes() const {
  return _axes;
}
const AxesSP<generic::Map<std::string, bool>>& BaseDiscreteSpace::increasing()
    const {
  return _increasing;
}
const instrument_interfaces::names::PortsSP BaseDiscreteSpace::knobs() const {
  instrument_interfaces::names::PortsSP knobs;
  for (const domains::CoupledLabelledDomainSP axis : *axes()) {
    for (const instrument_interfaces::names::InstrumentPortSP knob :
         *axis->labels()) {
      knobs->push_back(knob);
    }
  }
  return knobs;
}
void BaseDiscreteSpace::validate_unit_space_dimensionality_matches_knobs()
    const {
  if (_space->dimension() != _axes->size()) {
    throw std::invalid_argument(
        "Unit space dimensionality does not match number of knob axes.");
  }
}
void BaseDiscreteSpace::validate_knob_uniqueness() const {
  std::set<std::string> old_names;
  for (const domains::CoupledLabelledDomainSP& axis : *_axes) {
    std::set<std::string> new_names(
        axis->labels()->get_default_names()->begin(),
        axis->labels()->get_default_names()->end());
    assert(std::none_of(new_names.begin(),
                        new_names.end(),
                        [&](const std::string& name) {
                          return old_names.count(name);
                        }) &&
           "The default names must be unique.");
    old_names.insert(new_names.begin(), new_names.end());
  }
}
const int BaseDiscreteSpace::get_axis(
    const instrument_interfaces::names::InstrumentPortSP& knob) const {
  for (domains::CoupledLabelledDomainSP axis : *_axes) {
    if (axis->labels()->contains(knob)) {
      return _axes->index(axis);
    }
  }
  throw std::runtime_error("Knob " + knob->default_name() +
                           " not found in the axes.");
}
const domains::DomainSP BaseDiscreteSpace::get_domain(
    const instrument_interfaces::names::InstrumentPortSP& knob) const {
  int axis = get_axis(knob);
  return _axes->at(axis)->get_domain(knob);
}
const AxesSP<arrays::LabelledControlArray> BaseDiscreteSpace::get_projection(
    const AxesSP<instrument_interfaces::names::InstrumentPort>& projection)
    const {
  // Validate dimensionality
  if (projection->size() != _space->dimension()) {
    throw std::runtime_error(
        "The projection dimensionality must be less than or equal to the space "
        "dimensionality.");
  }

  // Get projection axes indices
  std::vector<int> projection_axes;
  for (const auto& knob : *projection) {
    projection_axes.push_back(get_axis(knob));
  }

  // Check uniqueness
  std::set<int> unique_axes(projection_axes.begin(), projection_axes.end());
  if (projection->size() != unique_axes.size()) {
    throw std::runtime_error("The projection axes must be unique.");
  }

  // Create unit projections
  auto unitprojections =
      _space->create_array(std::make_shared<Axes<int>>(projection_axes));

  std::vector<arrays::ControlArraySP> scaled_projections;
  for (size_t i = 0; i < unitprojections->size(); ++i) {
    auto unitprojection = unitprojections->at(i);
    auto knob           = projection->at(i);
    auto increasing_map = _increasing->at(i);

    auto        domain     = get_domain(knob);
    double      difference = domain->range();
    std::string index      = knob->instrument_facing_name();
    int         sign       = increasing_map->at(index) ? 1 : -1;
    double value = sign > 0 ? domain->lesser_bound() : domain->greater_bound();
    scaled_projections.push_back(*(*(*unitprojection * difference) * sign) +
                                 value);
  }
  generic::ListSP<arrays::LabelledControlArray> container;
  for (int i = 0; i <= projection_axes.size(); i++) {
    container->push_back(std::make_shared<arrays::LabelledControlArray>(
        scaled_projections.at(i), projection->at(i)));
  }
  return std::make_shared<Axes<arrays::LabelledControlArray>>(container);
}

}  // namespace falcon_core::math::discrete_spaces
CEREAL_REGISTER_TYPE(falcon_core::math::discrete_spaces::BaseDiscreteSpace)
CEREAL_REGISTER_POLYMORPHIC_RELATION(
    falcon_core::generic::Song,
    falcon_core::math::discrete_spaces::BaseDiscreteSpace)
